semver.py

```diff
@@ -125,33 +128,51 @@ def _asdict(self):
         ))
 
     def __eq__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
+            else:
+                return NotImplemented
         return _compare_by_keys(self._asdict(), _to_dict(other)) == 0
 
     def __ne__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
+            else:
+                return NotImplemented
         return _compare_by_keys(self._asdict(), _to_dict(other)) != 0
 
     def __lt__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
+            else:
+                return NotImplemented
         return _compare_by_keys(self._asdict(), _to_dict(other)) < 0
 
     def __le__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
+            else:
+                return NotImplemented
         return _compare_by_keys(self._asdict(), _to_dict(other)) <= 0
 
     def __gt__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
+            else:
+                return NotImplemented
         return _compare_by_keys(self._asdict(), _to_dict(other)) > 0
 
     def __ge__(self, other):
-        if not isinstance(other, (VersionInfo, dict)):
-            return NotImplemented
+        if not _is_comparable(other):
+            if PY2:
+                raise TypeError
```

Comment:
Hm, but why? NotImplemented existed for Python 2.x as well and there rules are the same for it.

Reply:
See previous build https://travis-ci.org/k-bx/python-semver/builds/404887965?utm_source=github_status&utm_medium=notification for Python 2.7 ([see job](https://travis-ci.org/k-bx/python-semver/jobs/404887966))

even if NotImplemented existed for Python 2.x returning it doesn't raise TypeError (contrary to Python 3.x) and comparison wasn't returning a correct result.

Reply:
No, NotImplemented in Py2.7 works in the same way: https://docs.python.org/2.7/library/constants.html#NotImplemented

Reply:
Ah, Python 2.x swallows them and returns always True or False...or False in our case.

Reply:
I think, it would be better to move that check and behaviour difference into tests. 

While it's good idea to keep same behaviour in both Pythons, Python 2.x world tells that every comparison end with boolean result, so we would looks here like a rule breaker.

Raising TypeError for Python 2.x also breaks NotImplemented behavior when if `a.__eq__(b)` fails, Python tries to do `b.__eq__(a)` - by raising exception we drops the second and our behaviour in both Pythons still remains inconsistent.